jsonpath: update error handling for undefined variable errors

normanchenn opened the pull request:
#### jsonpath: refactor walkAST and reorganize implementation

This commit refactors walkAST to a Validate method within the Path
interface to implement for all path types.

Release note: None

#### jsonpath: update error handling for undefined variable errors

This commit updates error handling logic for undefined variable errors
within JSONPath queries. Instead of silencing all errors into
jsonpathBoolUnknown in evalPredicate, we explicitly check if it's the
variable not found error. Additionally, this commit also allows for
assertion failures to be not silenced when passed to evalPredicate.

Closes: #143730
Fixes: #143692
Fixes: #144778
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144428)
<!-- Reviewable:end -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (last commit)

> Reviewed 2 of 2 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, 2 of 2 files at r4, 3 of 3 files at r5, 15 of 15 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-:-OO-GS255SNPsCG2GzEc:bqaf5gh)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-G5YU1hHg2A4Kajs4:-OO-G5YU1hHg2A4Kajs5:b-6bcl7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1616)):*
> ```
> SELECT * from jsonb_path_query('{"a": 10}', '$ ? (@.a < $value)', '{"value": 11}');
> ----
> {"a": 10}
> ```

Note that Postgres seems to do this variable check lazily, so that conditional short-circuiting prevents the error:
```
postgres=# SELECT * from jsonb_path_query('{"a": 10}', '$ ? ((@.a < 12) || (@.a < $value))');
 jsonb_path_query
------------------
 {"a": 10}
(1 row)
```
That being said, I think it's fine if we do the check more eagerly.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (dismissed):
Had a few more thoughts after I submitted.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-:-OO-HeOgDr6DjsNVeFEk:b-n4yrti)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @normanchenn)

___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-G5YU1hHg2A4Kajs4:-OO-HU-66j0LN7hSHyYb:b1ub9uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1616)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Note that Postgres seems to do this variable check lazily, so that conditional short-circuiting prevents the error:
```
postgres=# SELECT * from jsonb_path_query('{"a": 10}', '$ ? ((@.a < 12) || (@.a < $value))');
 jsonb_path_query
------------------
 {"a": 10}
(1 row)
```
That being said, I think it's fine if we do the check more eagerly.
</blockquote></details>

Let's also do tests with multiple variable references as well as multiple variables.
___
*[`pkg/util/jsonpath/jsonpath.go` line 30 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-H3ESA7YhrPH6OD06:-OO-H3ET6CpOLrjNRovs:b-ot8os7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/util/jsonpath/jsonpath.go#L30)):*
> ```Go
> 	}
> 	varsList := make([]string, 0, len(vars))
> 	for v := range vars {
> ```

Note: maps in go iterate in a nondeterministic order, so order in which these variables are checked is also not deterministic. That could lead to test flakes due to getting slightly different errors, so it might be worth sorting the variable names.


<!-- Sent from Reviewable.io -->

normanchenn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/util/jsonpath/jsonpath.go` line 30 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-H3ESA7YhrPH6OD06:-OO-pldZ5RkOV35sZiqB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/util/jsonpath/jsonpath.go#L30)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Note: maps in go iterate in a nondeterministic order, so order in which these variables are checked is also not deterministic. That could lead to test flakes due to getting slightly different errors, so it might be worth sorting the variable names.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-G5YU1hHg2A4Kajs4:-OO-sljY4vTDNSgkMohs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1616)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Let's also do tests with multiple variable references as well as multiple variables.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice! I have one suggestion.

> Reviewed 11 of 15 files at r6, 10 of 10 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-:-OOCwlH55W-6aXgO8IFt:bctmxe8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-G5YU1hHg2A4Kajs4:-OOCwd184YDWGsgnegDr:bpgtgzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1616)):*
<details><summary><i>Previously, normanchenn (Norman Chen) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Hm, I don't see a test case for lazy evaluation / short-circuiting logic that Drew pointed out.
___
*[`pkg/util/jsonpath/jsonpath.go` line 24 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OOCuV4a5f6PrWaCzzgY:-OOCuV4a5f6PrWaCzzgZ:b-7fzaub) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0392e93b54aa58e7fb6e69ea8578132ad31812b/pkg/util/jsonpath/jsonpath.go#L24)):*
> ```Go
> 
> // ValidateAndCollectVariables walks the Jsonpath AST and collects all variable
> // names. It returns an error if the AST is invalid (last not in array subscripts,
> ```

nit: are these two error conditions exhaustive for the AST being invalid? I.e. are there any other conditions that make the AST invalid? If so, include something like `, etc.)` into the comment.
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1573 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OOCvuxO1YwFVXvQH22E:-OOCvuxO1YwFVXvQH22F:b-9s7fwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0392e93b54aa58e7fb6e69ea8578132ad31812b/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1573)):*
> ```
> SELECT jsonb_path_query('{"a": 10, "b": 20}', '$ ? (@.a < $value && @.b > $value)');
> 
> # Postgres would outputs min, but we sort the variable names and then check for invalid variables.
> ```

nit: assuming that Postgres uses the variable name that is first mentioned in the error, we could match that behavior with simple extension - we could introduce a helper struct that would store the variable name as well as the "ordinal" of that variable which would equal the number of variables in the map when the given variable was parsed for the first time. Then we'd sort all variable names based on those ordinals, rather than lexicographically.


<!-- Sent from Reviewable.io -->

normanchenn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @michae2, and @yuzefovich)

___
*[`pkg/util/jsonpath/jsonpath.go` line 24 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OOCuV4a5f6PrWaCzzgY:-OONycbG83ruGEIQxGhN:b-26ddhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0392e93b54aa58e7fb6e69ea8578132ad31812b/pkg/util/jsonpath/jsonpath.go#L24)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: are these two error conditions exhaustive for the AST being invalid? I.e. are there any other conditions that make the AST invalid? If so, include something like `, etc.)` into the comment.
</blockquote></details>

These are the only ones we're currently checking.
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-G5YU1hHg2A4Kajs4:-OONm2RZArO0EK5r4D04:b-g8agds) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1616)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I don't see a test case for lazy evaluation / short-circuiting logic that Drew pointed out.
</blockquote></details>

I'll add in the test case for lazy evaluation, but I think implementing it would require some changes to the current implementation. We'd need to handle variable errors during evaluation instead of before, and to properly propagate variable errors, we might need to implement some kind of panic/catch mechanism since variable errors are always returned, while we suppress errors within filter/predicate evaluation. I can create an issue to track this.
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1573 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OOCvuxO1YwFVXvQH22E:-OOO1OVO8QienJEfWrWn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0392e93b54aa58e7fb6e69ea8578132ad31812b/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1573)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: assuming that Postgres uses the variable name that is first mentioned in the error, we could match that behavior with simple extension - we could introduce a helper struct that would store the variable name as well as the "ordinal" of that variable which would equal the number of variables in the map when the given variable was parsed for the first time. Then we'd sort all variable names based on those ordinals, rather than lexicographically.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

(You'll need to dismiss Drew's review - since it is blocking right now - before merging.)

> Reviewed 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-:-OORfYcU5yUWYYWiVoBr:b-4juext)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-G5YU1hHg2A4Kajs4:-OOReu-mBwJPDJFrPKZc:bojo3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1616)):*
<details><summary><i>Previously, normanchenn (Norman Chen) wrote…</i></summary><blockquote>

I'll add in the test case for lazy evaluation, but I think implementing it would require some changes to the current implementation. We'd need to handle variable errors during evaluation instead of before, and to properly propagate variable errors, we might need to implement some kind of panic/catch mechanism since variable errors are always returned, while we suppress errors within filter/predicate evaluation. I can create an issue to track this.
</blockquote></details>

Ack. I think it'd be good to create a separate issue with "docs known limitation" label so that we document this difference (and also mention the issue number in the comment in the logic test).


<!-- Sent from Reviewable.io -->

normanchenn reviewed (commented):
> Dismissed @DrewKimball from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @michae2, and @yuzefovich)

___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OO-G5YU1hHg2A4Kajs4:-OOSwo4z2vdcOmekfc49:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a66e771d928f0eaf6ef0afa5084cd96ba858aca/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1616)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ack. I think it'd be good to create a separate issue with "docs known limitation" label so that we document this difference (and also mention the issue number in the comment in the logic test).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

normanchenn committed:
jsonpath: refactor walkAST and reorganize implementation

This commit refactors walkAST to a Validate method within the Path
interface to implement for all path types.

Release note: None

normanchenn committed:
jsonpath: update error handling for undefined variable errors

This commit updates error handling logic for undefined variable errors
within JSONPath queries. Instead of silencing all errors into
jsonpathBoolUnknown in evalPredicate, we explicitly check if it's the
variable not found error. Additionally, this commit also allows for
assertion failures to be not silenced when passed to evalPredicate.

Release note: None

normanchenn commented:
After discussing with @mgartner, I've changed this PR to check the variables during evaluation, and fixing an issue with how we handle errors within predicates. This should resolve #143692 and #144778.

mgartner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a couple minor suggestions

> Reviewed 2 of 15 files at r6, 13 of 13 files at r10, 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-:-OOU4msK71TRjKru2u2c:b-t0e3a0)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @michae2)

___
*[`pkg/util/jsonpath/eval/operation.go` line 267 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OOU3DVr2B-FeAlTw8L3:-OOU3DVr2B-FeAlTw8L4:bv1z657) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec6ee18df5d28c074221615c9aa4d647f947805b/pkg/util/jsonpath/eval/operation.go#L267)):*
> ```Go
> 	left, err := ctx.evalAndUnwrapResult(op.Left, jsonValue, true /* unwrap */)
> 	if err != nil {
> 		if errors.HasAssertionFailure(err) || errors.Is(err, errVariableNotFound) {
> ```

nit: make this a helper function like `!isIgnorableError(err)` and then you can simplify to:

```go
if !isIgnorableError(err) {
    return jsonpathBoolUnknown, err
}
if err != nil || left == nil {
    return jsonpathBoolUnknown, nil
}
```
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 1642 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OOU4eRHEaL4NdOcZyZ8:-OOU4eRHEaL4NdOcZyZ9:b-89q77x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec6ee18df5d28c074221615c9aa4d647f947805b/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L1642)):*
> ```
> SELECT jsonb_path_query('{"a": 10}', '$ ? ((@.a < 12) || (@.a < $value))');
> ----
> {"a": 10}
> ```

Cool!
___
*[`pkg/util/jsonpath/eval/scalar.go` line 27 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/144428#-OOU2WLb4nAQouYyPUSq:-OOU2WLb4nAQouYyPUSr:b-de82vl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec6ee18df5d28c074221615c9aa4d647f947805b/pkg/util/jsonpath/eval/scalar.go#L27)):*
> ```Go
> 		}
> 		if val == nil {
> 			return nil, errors.Wrapf(errVariableNotFound, "could not find jsonpath variable %q", scalar.Variable)
> ```

An issue with `errors.Wrapf` is that it attaches a stack trace to the error, which is expensive to do and isn't always necessary. I don't think it is necessary here. I think we can use `errors.Mark` instead. I think you can make the `errVariableNotFound` simply an empty struct `type errVariableNotFound struct{}` and create the pg error here:

```go
return nil, errors.Mark(pgerror.Newf(pgcode.UndefinedObject, "could not find jsonpath variable %q", scalar.Variable), &errVariableNotFound{})
```


<!-- Sent from Reviewable.io -->

normanchenn commented:
TFTRs!

bors r+
